docs/RFCS/20220112_pebble_sstable_format_versions.md

```diff
@@ -0,0 +1,242 @@
+
+In the context of a distributed system such as Cockroach, certain SSTable
+features are backwards incompatible (e.g. the block property collection and
+filtering feature extends the RocksDBv2 SSTable block index format to encoded
+various block properties, which is a breaking change). Participants must
+_first_ ensure that their stores have the code-level features available to read
+and write these newer SSTables (indicated by Pebble's format major version).
+Once all stores agree that they are running the minimum Pebble format major
+version and will not roll back (e.g. Cockroach cluster version finalization),
+SSTables can be written and read using more recent table formats. The Pebble
+"format major version" and "table format version" are therefore no longer
+independent - the former implies an upper bound on the latter.
+
+As a real-world example of the need for the above, consider two Cockroach nodes
+each with a Pebble store, one at version A, the other at version B (version A
+(newer) > B (older)). Store A constructs an SSTable for an external backup
+containing a newer block index format (for block property collection). This
+SSTable is then imported in to store B. Store B fails to read the SSTable as it
+is not running with a format major version recent enough make sense of the
+newer index format. The two stores require a method for agreeing on a minimum
+supported table format.
+
+The remainder of this document outlines a new table format for Pebble. This new
+table format will be used for new table-level features such as block properties
+and range keys (see
+[#1339](https://github.com/cockroachdb/pebble/issues/1339)), but also for
+backporting table-level features from RocksDB that would be useful to Pebble
+(e.g. version 3 avoids encoding sequence numbers in the index, and version 4
+uses delta encoding for the block offsets in the index, both of which are
+useful for Pebble).
+
+# Technical design
+
+## Pebble magic number
+
+The last 8 bytes of an SSTable is referred to as the "magic number".
+
+LevelDB uses the first 8 bytes of the SHA1 hash of the string
+`http://code.google.com/p/leveldb/` for the magic number.
+
+RocksDB uses its own magic number, which indicates the use of a slightly
+different table layout - the footer (the name for the end of an SSTable) is
+slightly larger to accommodate a 32-bit version number and 8 bits for a
+checksum type to be used for all blocks in the SSTable.
+
+A new magic number will be introduced for Pebble. Similar to LevelDB, the first
+8 bytes of the SHA1 hash of the string "github.com/cockroachdb/pebble" will be
+used:
+
+```
+SHA1('github.com/cockroachdb/pebble') = \x69\xda\xf0\x0e\x5c\x1d\x47\x82
+```
+
+## Pebble version scheme
+
+Tables with a Pebble magic number will use a dedicated versioning scheme,
+starting with version `1`. No new versions other than version `2` will be
+supported for tables containing the RocksDB magic number.
```

Comment:
This diverged from the initial proposal. I think this is much simpler if we can assume we don't need to support other RocksDB versions (seems reasonable?).